napari/plugins/_npe2.py

```diff
@@ -125,19 +144,132 @@ def get_widget_contribution(
     return None
 
 
-def populate_qmenu(menu: QMenu, menu_key: str):
-    """Populate `menu` from a `menu_key` offering in the manifest."""
-    # TODO: declare somewhere what menu_keys are valid.
+def _build_menus(entries: Iterable[MenuEntry]) -> List[MenuItem]:
+    """Build napari native menus from npe2 menu entries.
+
+    Parameters
+    ----------
+    entries : iterable of npe2.manifest.contributions.MenuItem
+        npe2 menu entries to build from.
+
+    Returns
+    -------
+    menu_item : list of napari.components.menu.MenuItem
+        Built napari native menu items.
+
+    Raises
+    ------
+    TypeError
+        If any menu entry is not a menu command or submenu.
+    """
+    return [build_menu_item(entry) for entry in entries]
+
+
+@functools.cache
```

Comment:
also love, but also 3.9+ only... still need to use `lru_cache` for now.  Also, do we need to add a `build_submenu.cache_clear()` call to the `_on_plugin_enablement_change` below?  (otherwise, if a plugin is installed via the builtin plugin installer... it won't be reflected in the menus right?)

Reply:
the cache will still work fine if new plugins are added but you're right, if one is removed the function will erroneously have the submenu built in its cache